### PR TITLE
#6/add google storage

### PR DESCRIPTION
**Pull request by wojtek-krysiak**
Adding Google Storage to the upload options

### CODE REVIEW

File: src/features/upload-file/providers/gcp-provider.ts [hunk 1]
```diff
@@ -0,0 +1,78 @@
+import { UploadedFile } from 'admin-bro'
```
Comment (wojtek-krysiak) [inline src/features/upload-file/providers/gcp-provider.ts:1]: this is an actual gcloud integration

File: tsconfig.json [hunk 2]
```diff
@@ -13,7 +13,10 @@
     "noImplicitThis": true,
     "moduleResolution": "node",
     "module": "commonjs",
-    "baseUrl": "."
+    "baseUrl": ".",
+    "paths": {
+      "react": ["node_modules/@types/react"]
```
Comment (wojtek-krysiak) [inline tsconfig.json:18]: this fixes the issue with linking packages (`yarn link`) having the same type definitions

File: package.json [hunk 3]
```diff
@@ -8,7 +8,9 @@
   "license": "MIT",
   "scripts": {
     "release": "semantic-release",
-    "build": "tsc",
+    "build": "yarn tsc",
+    "clean": "rm -rf build && mkdir build && rm -fR types && mkdir types",
```
Comment (plorencrstit) [inline package.json:12]: once `-rf`, other time `-fR`

File: src/features/upload-file/property-custom.type.ts [hunk 4]
```diff
@@ -16,7 +14,7 @@ type PropertyCustom = {
   defaultBucket: string,
   mimeTypes?: Array<MimeType>,
   maxSize?: number,
-  provider: keyof Required<Exclude<UploadOptions['provider'], BaseProvider>> | 'base'
+  provider: string
```
Comment (plorencrstit) [inline src/features/upload-file/property-custom.type.ts:17]: why can not be some enum?

File: src/features/upload-file/providers/gcp-provider.ts [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+      // eslint-disable-next-line
+      GCPStorage = require('@google-cloud/storage').Storage
+    } catch (error) {
+      throw new Error(
+        'You have to install "@google-cloud/storage" in order to run this plugin with GCP',
+      )
+    }
+    // // this check is needed because option expires can be `0`
+    this.expires = typeof options.expires === 'undefined'
+      ? 86400
```
Comment (plorencrstit) [inline src/features/upload-file/providers/gcp-provider.ts:45]: [OPT] could be some variable which says how much time it is

File: src/features/upload-file/upload-file.feature.ts [hunk 6]
```diff
@@ -30,6 +28,10 @@ const uploadFileFeature = (config: UploadOptions): FeatureType => {
     const options = (provider as ProviderOptions).aws
     adapter = new AWSProvider(options)
     providerName = 'aws'
+  } else if (provider && (provider as ProviderOptions).gcp) {
+    const options = (provider as ProviderOptions).gcp
+    adapter = new GCPProvider(options)
+    providerName = 'gcp'
```
Comment (plorencrstit) [inline src/features/upload-file/upload-file.feature.ts:34]: It should be enum

**Review by plorencrstit — APPROVED** [review]

File: example-app/package.json [hunk 7]
```diff
@@ -16,12 +17,14 @@
...
     "nodemon": "^2.0.4",
     "ts-node": "^8.10.2",
-    "typescript": "^3.9.7"
+    "typescript": "^3.9.7",
+    "wait-on": "^5.2.0"
   },
   "dependencies": {
     "@admin-bro/express": "^3.0.0",
     "@admin-bro/typeorm": "1.2.0",
     "@admin-bro/upload": "v1.0.0-beta.3",
```
[the remark this answers is not in the document]
Comment (SimonB407) [reply example-app/package.json:26]: TBH I really don't like adding dependency to library that example is for, it messes up example app experience. For example you want to check, if after some changes example app works, you have to always remember to change it from npm library to local changes. I know it messes up ts builds (weird dir structure) but in my opinion it's more convenient, to just depend on code from local, than from npm package.
```suggestion
    "@admin-bro/upload": "../src",
```

File: src/features/upload-file/build-path.ts [hunk 8]
```diff
@@ -18,6 +18,9 @@ const buildRemotePath = (
   if (!record.id()) {
     throw new Error('You cannot upload file for not persisted record. Save record first')
   }
+  if (!file.name) {
+    throw new Error('Server could not have verified the file name')
```
Comment (SimonB407) [inline src/features/upload-file/build-path.ts:22]:
```suggestion
    throw new Error('Server could not verify the file name')
```

File: src/features/upload-file/upload-options.type.ts [hunk 9]
```diff
@@ -15,6 +16,8 @@ type UploadOptions = {
   provider: {
     /** AWS Credentials */
     aws?: AWSOptions,
+    /** AWS Credentials */
```
Comment (SimonB407) [inline src/features/upload-file/upload-options.type.ts:19]:
```suggestion
    /** AGCP Credentials */
```

File: src/features/upload-file/property-custom.type.ts (hunk 4 above)
Comment (JonaszJestem) [reply src/features/upload-file/property-custom.type.ts:17]: :+1: or just `'gcp' | 'aws' | 'local'`

File: src/features/upload-file/providers/gcp-provider.ts [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+
+  public expires: number
+
+  constructor(options: GCPOptions) {
+    super(options.bucket)
+
+    let GCPStorage: typeof Storage
+    try {
+      // eslint-disable-next-line
+      GCPStorage = require('@google-cloud/storage').Storage
```
Comment (JonaszJestem) [inline src/features/upload-file/providers/gcp-provider.ts:37]: Optional: doesn't matter in this case but this should work too and will eliminate eslint warning
```suggestion
      import('@google-cloud/storage').then(({Storage}) => { this.GCPStorage = new Storage() }).catch([...])
```

**Review by JonaszJestem — APPROVED** [review]

File: src/features/upload-file/build-path.ts (hunk 8 above)
Comment (wojtek-krysiak) [reply src/features/upload-file/build-path.ts:22]: thanks - will change

File: src/features/upload-file/upload-options.type.ts (hunk 9 above)
Comment (wojtek-krysiak) [reply src/features/upload-file/upload-options.type.ts:19]: thanks

**Review by SimonB407 — APPROVED** [review]

**Review by JonaszJestem — APPROVED** [review]

**Pull request merged**